ENH: Refactor the model so it is compatible with DIPY's interface

oesteban opened the pull request:
The strategy is to initiate the new `eddymotion.model.dipy`, where we test out DIPY "extensions" which potentially will be added to their codebase. In it, a new GP model is added.

Our base module is then updated to handle the GP model as a standard DIPY model.

cc/ @arokem

oesteban committed:
wip: write the model looking into DIPY

arokem commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,150 @@
+
+from sklearn.gaussian_process import GaussianProcessRegressor
+from dipy.reconst.base import ReconstModel
+
```
```suggestion
from dipy.reconst.multi_voxel import multi_voxel_fit
```

arokem commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,150 @@
+        self.kernel_model = kernel_model
+        self.mask = mask
+
+    def fit(self, data, mask=None, random_state=0):
```
```suggestion
    @multi_voxel_fit
    def fit(self, data, mask=None, random_state=0, **kwargs):
```

arokem replied:
The additional kwargs let you control the parallelization.

arokem reviewed (commented):
Looking good. This is how you implement parallelization as in https://github.com/dipy/dipy/pull/2593

oesteban committed:
Apply suggestions from code review

Co-authored-by: Ariel Rokem <arokem@gmail.com>

oesteban committed:
enh: deep revision

arokem commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,209 @@
+        )
+
+    @multi_voxel_fit
+    def multi_fit(self, data_thres, mask=None, **kwargs):
```
What is the purpose of this method? Also, looks like the mask doesn't get used below.

oesteban replied:
I separated the decorated fit because if I interpret the `multi_voxel_fit` decorator properly, an array with shape (n_parameters, n_voxels) is expected as the output of the decorated function, so that then parameters can be stacked/manipulated and used in prediction.

In our case, instead, we get a trained GaussianProcessRegressor object instance. We can probably extract the parameters that describe it, but I don't know how to do that at this point.

So I moved it out to punt the implementation at a later time when everything is working for a single thread single voxel fit.

WDTY?

arokem reviewed (commented):
This is great! With a bit more unit testing, I think that this will be easy to port upstream to DIPY down the line. 

For now, I am just wondering about the separation between `fit` and `multi_fit`. What's the purpose of that?

oesteban commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,209 @@
+        raise RuntimeError("Model is not yet fitted.")
+
+    # Extract orientations from gtab, and highly likely, the b-value too.
+    return model._gpr.predict(gtab, return_std=False)
```
@jhlegarreta I'm going to walk back my words and now I'm confident that the reshaping (w/ or w/o mask) should happen at a higher level of the class hierarchy. Basically, all models we have so far need to do it, so let's not re-implement the same operation for every model.

WDTY?

jhlegarreta replied:
Seems reasonable. Did not have time to have a look yesterday, so I am probably missing some context, but if you now have a clearer picture across all models and believe so, I am fine with it.

oesteban committed:
enh: finalizing PR and formatting

oesteban committed:
enh: bring utilities into the new dipy module

oesteban committed:
enh: revise pairwise angle computation to allow X and Y

oesteban commented:
@jhlegarreta are you comfortable merging this now? That will allow me to make further steps before tomorrow's meeting.

I believe with my last commit, things are in a very good state to be set up as a base for the GP.

jhlegarreta commented:
> @jhlegarreta are you comfortable merging this now? That will allow me to make further steps before tomorrow's meeting.

A few comments:
- I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository: https://github.com/nipreps/eddymotion/pull/212.
- The intermediate commits (removing the `viz` module, adding the covariance models, etc.) make the review  more difficult. Rebasing on the fork `master` should have been enough to only leave the relevant commits and changes.
- Moving things (gradient angle computation, covariance computation, etc.) to a new `.dipy` module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

But I guess my comments are of little value if this needs to be merged now. @oesteban let me know if you really believe that addressing my comments is unfeasible/not useful, and will merge.

oesteban commented:
> * I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository:

My bad, I missed that one. Will work on it tomorrow, merge and then downstream here.

> The intermediate commits (removing the viz module, adding the covariance models, etc.) make the review more difficult. Rebasing on the fork master should have been enough to only leave the relevant commits and changes.

I merged upstream/main into my branch precisely to make it easier to review, but yes, your branch would require a rebase to make the diff easier.

oesteban commented:
Failed to answer the last one:

> Moving things (gradient angle computation, covariance computation, etc.) to a new .dipy module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

Should this new model be upstreamed to DIPY, these functions must also be propagated there or call equivalent and pre-existing implementations. I do not foresee many more models being upstreamed to DIPY, so it feels okay if all these items are kept together without proliferating small modules. Regarding debugging and testing, I'm not so clear that having these functions scattered in small modules is helpful. On the contrary, it may be easier if the stack trace is in a single file.

oesteban commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,417 @@
+
+
+def compute_pairwise_angles(
+    gtab_X: GradientTable,
+    gtab_Y: GradientTable | None = None,
```
Since this will probably make it into dipy one day, my point of having these being ndarrays does not stand. Better use the gradient table object, which gives us two goodies:

- implicit checking of dimensionality (and therefore, dropped checks within the implementation)
- access to the b-values (which will come handy when we support multi-shell kernels).

jhlegarreta replied:
:+1: Resolving.

oesteban commented on src/eddymotion/model/dipy.py:
```diff
@@ -0,0 +1,417 @@
+    bvecs_X = gtab_X.bvecs.T
+    bvecs_X = np.array(bvecs_X) / np.linalg.norm(bvecs_X, axis=0)
+
+    if gtab_Y is None:
+        bvecs_Y = bvecs_X
```
Support the old use case (auto-covariance) if the second vector is not provided.

jhlegarreta replied:
:+1: Resolving. I'll add a test case for this in a separate PR.

oesteban commented:
>     * I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository: [ENH: Make grad vec angle computation method accept two input parameters nipreps/eddymotion#212](https://github.com/nipreps/eddymotion/pull/212).

Addressed - #212 is now merged and I have updated downstream. Apologies for the oversight.
 
>     * The intermediate commits (removing the `viz` module, adding the covariance models, etc.) make the review  more difficult. Rebasing on the fork `master` should have been enough to only leave the relevant commits and changes.

I have rebased your branch and this PR - diff should be clean now.


>     * Moving things (gradient angle computation, covariance computation, etc.) to a new `.dipy` module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

This is the only outstanding point, which I hope I made sense responding above - let's chat later :)

jhlegarreta commented:
OK, I'm merging this after our conversation, and we move forward from there. Thanks Óscar.
